crates/bevy_scene/src/dynamic_scene_builder.rs

```diff
@@ -0,0 +1,230 @@
+
+    /// Prepare a builder that will extract entities and their component from the given [`World`].
+    /// Only components registered in the given [`AppTypeRegistry`] will be extracted.
+    pub fn from_world_with_type_registry(world: &'w World, type_registry: AppTypeRegistry) -> Self {
+        Self {
+            scene: default(),
+            type_registry,
+            world,
+        }
+    }
+
+    /// Consume the builder, producing a [`DynamicScene`].
+    pub fn build(self) -> DynamicScene {
+        DynamicScene {
+            entities: self.scene.into_values().collect(),
+        }
+    }
+
+    /// Extract one entity from the builder's [`World`].
+    pub fn extract_entity(&mut self, entity: Entity) -> &mut Self {
+        self.extract_entities(std::iter::once(entity))
+    }
+
+    /// Extract entities from the builder's [`World`].
+    ///
+    /// Extracting entities can be used to extract entities from a query:
+    /// ```
+    /// # use bevy_scene::DynamicSceneBuilder;
+    /// # use bevy_app::AppTypeRegistry;
+    /// # use bevy_ecs::{
+    /// #     component::Component, prelude::Entity, query::With, reflect::ReflectComponent, world::World,
+    /// # };
+    /// # use bevy_reflect::Reflect;
+    /// #[derive(Component, Default, Reflect)]
+    /// #[reflect(Component)]
+    /// struct MyComponent;
+    ///
+    /// # let mut world = World::default();
+    /// # world.init_resource::<AppTypeRegistry>();
+    /// # let _entity = world.spawn(MyComponent).id();
+    /// let mut query = world.query_filtered::<Entity, With<MyComponent>>();
+    ///
+    /// let mut builder = DynamicSceneBuilder::from_world(&world);
+    /// builder.extract_entities(query.iter(&world));
+    /// let scene = builder.build();
+    /// ```
+    pub fn extract_entities(&mut self, entities: impl Iterator<Item = Entity>) -> &mut Self {
+        let type_registry = self.type_registry.read();
+
+        for entity in entities {
+            if !self.scene.contains_key(&entity.id()) {
+                let mut entry = DynamicEntity {
+                    entity: entity.id(),
+                    components: Vec::new(),
+                };
+
+                for component_id in self.world.entity(entity).archetype().components() {
+                    let reflect_component = self
+                        .world
+                        .components()
+                        .get_info(component_id)
+                        .and_then(|info| type_registry.get(info.type_id().unwrap()))
+                        .and_then(|registration| registration.data::<ReflectComponent>());
+
+                    if let Some(reflect_component) = reflect_component {
+                        if let Some(component) = reflect_component.reflect(self.world, entity) {
+                            entry.components.push(component.clone_value());
+                        }
+                    }
+                }
+                self.scene.insert(entity.id(), entry);
+            }
```

Comment:
Nit: To reduce nesting, you could use a `continue`:
```suggestion
            if self.scene.contains_key(&entity.id()) {
                continue;
            }

            let mut entry = DynamicEntity {
                entity: entity.id(),
                components: Vec::new(),
            };

            for component_id in self.world.entity(entity).archetype().components() {
                let reflect_component = self
                    .world
                    .components()
                    .get_info(component_id)
                    .and_then(|info| type_registry.get(info.type_id().unwrap()))
                    .and_then(|registration| registration.data::<ReflectComponent>());

                if let Some(reflect_component) = reflect_component {
                    if let Some(component) = reflect_component.reflect(self.world, entity) {
                        entry.components.push(component.clone_value());
                    }
                }
            }

            self.scene.insert(entity.id(), entry);
```